File: src/mbs/application/article/get.py
from datetime import datetime
from typing import Protocol

from pydantic import BaseModel

from mbs.adapters.database.models import File
from mbs.application.common.article_gateway import ArticleReader, ArticleWriter, ArticleRater
from mbs.application.common.exceptions import NotFound, Forbidden, Unauthorized
from mbs.application.common.id_provider import IdProvider
from mbs.application.common.interactor import Interactor

import mbs.domain.exceptions as domain_exceptions
from mbs.application.common.storage_gateway import StorageAccessLinkMaker
from mbs.application.common.tag_gateway import TagReader
from mbs.domain.models import ArticleState, ArticleId, FileId, UserId
from mbs.domain.services.access import AccessService
from mbs.domain.services.article import ArticleService


class ArticleResult(BaseModel):
    id: ArticleId
    title: str
    content: str
    poster: FileId | None
    views: int
    likes: int
    tags: list[str]
    is_rated: bool
    state: ArticleState
    files: list['FileItem']
    author_id: UserId

    created_at: datetime
    updated_at: datetime | None


class FileItem(BaseModel):
    id: FileId
    url: str
    filename: str
    content_type: str
    created_at: datetime


class ArticleGateway(ArticleReader, ArticleWriter, ArticleRater, Protocol):
    pass


class GetArticle(Interactor[ArticleId, ArticleResult]):

    def __init__(
            self,
            article_gateway: ArticleGateway,
            article_service: ArticleService,
            storage_access_link_maker: StorageAccessLinkMaker,
            tag_reader: TagReader,
            access_service: AccessService,
            id_provider: IdProvider,
    ):
        self._article_gateway = article_gateway
        self._article_service = article_service
        self._storage_access_link_maker = storage_access_link_maker
        self._tag_reader = tag_reader
        self._access_service = access_service
        self._id_provider = id_provider

    async def __call__(self, data: ArticleId) -> ArticleResult:
        article, files = await self._article_gateway.get_article_with_files(data)
        if not article:
            raise NotFound("Публикация не найдена")

        try:
            self._access_service.ensure_can_get_article(
                is_auth=self._id_provider.is_auth(),
                user_id=self._id_provider.user_id(),
                article_author_id=article.owner_id,
                permissions=self._id_provider.permissions(),
                user_state=self._id_provider.user_state(),
                article_state=article.state,
            )
        except domain_exceptions.AuthenticationError as error:
            raise Unauthorized(str(error))
        except domain_exceptions.AccessDenied as error:
            raise Forbidden(str(error))

        updated_article = self._article_service.update_article(
            article,
            title=article.title,
            content=article.content,
            state=article.state,
            views=article.views + 1,  # Could there be problems due to a non-atomic operation? 🤔
            poster=article.poster_id,
            tags=article.tags,
        )

        # todo gather
        await self._article_gateway.save_article(updated_article)
        is_rated = await self._article_gateway.is_article_rated(article.id, self._id_provider.user_id())

        return ArticleResult(
            id=article.id,
            title=article.title,
            content=article.content,
            poster=article.poster_id,
            views=article.views,
            likes=article.likes,
            tags=article.tags,
            is_rated=is_rated,
            state=article.state,
            files=[
                FileItem(
                    id=file.id,
                    url=await self._storage_access_link_maker.make_article_download_link(  # todo gather
                        article_id=article.id,
                        file_id=file.id
                    ),
                    filename=file.filename,
                    content_type=file.content_type,
                    created_at=file.created_at
                ) for file in files
            ],
            author_id=article.author_id,
            created_at=article.created_at,
            updated_at=article.updated_at,
        )

File: src/application/article/get.rs
use crate::application::common::{
    article_gateway::ArticleGateway,
    error::AppError,
    file_map_gateway::FileMapReader,
    file_storage_gateway::FileStorageLinker,
    id_provider::IdProvider,
    interactor::Interactor
};
use crate::domain::models::rate_state::RateState;
use crate::domain::{
    models::{
        article::ArticleId,
        article_state::ArticleState,
        file::FileId,
        tag::Tag,
        user_id::UserId
    },
    services::access::ensure_can_get_article
};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use utoipa::{IntoParams, ToSchema};

#[derive(Deserialize, IntoParams)]
pub struct GetArticleInput {
    #[param(example = uuid::Uuid::new_v4, value_type=uuid::Uuid)]
    pub id: ArticleId
}

#[derive(Serialize, ToSchema)]
pub struct ArticleFile {
    #[schema(example = uuid::Uuid::new_v4, value_type=uuid::Uuid)]
    pub id: FileId,
    #[schema(example = "example.png")]
    pub filename: String,
    #[schema(example = "image/png")]
    pub content_type: String,
    #[schema(example = "https://s3.example.com/article-assets/987e9dc9-d84c-4ba7-837f-db755a0fdc55/80bbc0bc-4064-420a-b4ed-4f94b4575321")]
    pub url: String,
    
    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[derive(Serialize, ToSchema)]
pub struct GetArticleOutput {
    #[schema(example = "Super rust tips")]
    pub title: String,
    #[schema(example = uuid::Uuid::new_v4, value_type=Option<uuid::Uuid>)]
    pub poster: Option<FileId>,
    #[schema(example = "In this article, we will explore some of the best practices in Rust programming...")]
    pub content: String, // todo: server rendering 
    pub state: ArticleState,
    #[schema(example = 100)]
    pub views: u32,
    #[schema(example = -5)]
    pub rating: i64,
    #[schema(example = uuid::Uuid::new_v4, value_type=uuid::Uuid)]
    pub author_id: UserId,
    #[schema(example = json!(vec!["rust", "programming", "tips"]), value_type = Vec<String>)]
    pub tags: Vec<Tag>,
    
    pub self_rate: RateState,
    pub files: Vec<ArticleFile>,

    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>
}

pub struct GetArticle<'interactor> {
    pub id_provider: Box<dyn IdProvider>,
    pub article_gateway: &'interactor dyn ArticleGateway,
    pub file_map_reader: &'interactor dyn FileMapReader,
    pub file_storage_linker: &'interactor dyn FileStorageLinker,
}

impl Interactor<GetArticleInput, GetArticleOutput> for GetArticle<'_> {
    async fn execute(&self, input: GetArticleInput) -> Result<GetArticleOutput, AppError> {
        let article = match self.article_gateway.get_article(&input.id).await? {
            Some(article) => article,
            None => return Err(AppError::NotFound("id".into()))
        };

        ensure_can_get_article(
            self.id_provider.permissions(),
            self.id_provider.user_state(),
            &article.state,
            &article.author_id,
            self.id_provider.user_id(),
        )?;
        
        let (self_rate, files, inc_res) = tokio::join!(
            self.article_gateway.user_rate_state(&article.id, self.id_provider.user_id()),
            self.file_map_reader.get_article_files(&article.id),
            self.article_gateway.increment_article_views(&article.id)
        );
        
        inc_res?;
        
        Ok(GetArticleOutput {
            title: article.title,
            poster: article.poster,
            content: article.content,
            state: article.state,
            views: article.views,
            rating: article.rating,
            author_id: article.author_id,
            tags: article.tags,

            self_rate: self_rate?,
            files: files?.into_iter().filter(|f| f.is_uploaded).map(|file| {
                ArticleFile {
                    id: file.id,
                    filename: file.filename,
                    content_type: file.content_type,
                    url: self.file_storage_linker.download_link(&article.id, &file.id),
                    created_at: file.created_at,
                    updated_at: file.updated_at
                }
            }).collect(),
            
            created_at: article.created_at,
            updated_at: article.updated_at
        })
    }
}
